### ISSUE TITLE
Allow manually changing canvas size

### ISSUE THREAD

**Issue by vaunchag** [ksnip/kImageAnnotator#92]
FR:
Enhance the crop tool to permit the user not only to crop, but also to extend the canvas of the image (in any choosen direction, as per user preference).

Tool name may be changed for discoverability of this feature.

Initial reason is to control results (specially the extra margin) when doing out of border annotations (see #90).

Besides, even without out of border annotations involved, it is a nice to have feature: being able to directly add external borders (for aesthetics, for controlling the final aspect ratio, ...). 

Also related with #91 when implemented (the extension should fill with that chosen color).

**Comment by DamirPorobic** [comment]
I would prefer to have a separate option for this, leaving the crop as it is and adding new option called Extend Canvas area or Extend Scene like it's called in Qt terms.

**Comment by c33s** [comment]
i only can give feedback from photoshop perspective, it is really handy to have it in the crop-tool.
my suggestion would be to add it as extra "tool" but allow to enable the function for the crop tool.

**Comment by DamirPorobic** [comment]
In Photoshop it's part of the crop tool? I try to align to well known tools, they have usually more experienced people around UI design. It is even simpler to allow it in the crop tool because it's simple removing a restriction

**Comment by c33s** [comment]
![2OU87PLf1V](https://user-images.githubusercontent.com/649209/102614316-4e23b580-4134-11eb-92da-6c38404896ed.gif)

**Comment by DamirPorobic** [comment]
Interesting, then lets go with the crop.

**Comment by vaunchag** [comment]
Sounds great! Originally, I proposed to rename:
    crop
to something like
    crop / extend,
to enhance discoverability. But if Photoshop calls all this simply "crop",
I guess it can remain with that name.

Just in case, I also recall the related #91: if easy to implement, it's
quite complementary to this (and useful for other reasons, explained there).

>

**Comment by DamirPorobic** [comment]
I'm currently looking into this and I'm not all to happy with the combined Crop and Extend, the behavior is kind of strange that sometimes the images is cut of and sometimes not. I'll try now with two separate controls, one that can crop the background image, which cannot be extended beyond the background image (this is what we have already) and one extend or resize canvas that can increase the canvas size but cannot be shrinkt below the background image size. Later on we might add here a selection for the canvas fill.

**Comment by DamirPorobic** [comment]
The name could be something like "Adjust Canvas", that might allow doing other canvas operations here

**Comment by vaunchag** [comment]
I don't see well why you find it strange: one just adjusts freely the
region to get the desired view (as it works in crop tool in Photoshop).
While dragging, one should have visible at any moment the original frame,
just to have that reference on what he is doing.
    In fact, sometimes both things (crop/expand) may be needed at the same
time: some side has to be croped (too much was captured there), another
side has to be extended (e.g., in order leave room for annotations and
controlling the extra margin they will have there).
    In any case, if you finally prefer implementing this with 2 separate
tools, it is also good (the previous scenario can also be done in 2
separate steps).

**Comment by DamirPorobic** [comment]
Take for example an Image that you have captured 500x500 px, you open crop, cut one half of it of, the visible image is now 250x500 px and the canvas has the same size. You open the crop tool again, extend the canvas size to be again 500x500, but you don't get the one half of the image back.

If you don't cut the half of of the image, you can't see that anything happened, there is not border to show you what part of image is going to get into final image when you export it. 

Those scenarios confused even me while I was testing the feature.

I think keeping it in two tools might be more clear to users what they can do here and what not. That you need to use both tools is probably going to be a rare scenario and yes, might require opening two tools.

**Comment by c33s** [comment]
can we have a config for this? only because it's confusing for one it can be helpful for the other

**Comment by DamirPorobic** [comment]
A config is not going to work for this as the change would affect several places, making the code harder to maintain. I think that the solution I'm working on could work for for everyone. Basically the crop functionality is just used to remove parts of the background image like before, no change here. The Modify Canvas option can be used to resize the canvas which is currently by default `background Image` + `all items` + `image effects`, like this (existing functionality, just now visible):
![image](https://user-images.githubusercontent.com/12796236/103309326-1e6a8c80-4a15-11eb-8151-90d19ef3d07e.png)

By default it can be extended to any size you desire and is restricted to shrink only to the size of the background image: 
![image](https://user-images.githubusercontent.com/12796236/103309390-41953c00-4a15-11eb-9d93-3151ed5d0440.png)

Though this restriction can be disabled with the checkbox, then you can shrink like you want. 

What you finally see in this rect, will be exported when saving the image, even when the canvas is smaller then the image, like this, only the not dimmed part will be exported:

![image](https://user-images.githubusercontent.com/12796236/103309599-ce3ffa00-4a15-11eb-9c4f-7b83ca191f2c.png)

Hope this satisfies most wishes, I can't see any use case that could easier (with less operations) be achieved with a combined crop tool, if you wanted to cut a piece out on the right side and then increase the canvas in that direction, you would have need to use the combined crop tool twice. If you want to hide the right side and extend the canvas to the left side, this can be done now with just the modify canvas tool, opening it once. 

The thing is not finished, I'm still working on it and need to test it a bit further, might be finished tomorrow.

**Comment by vaunchag** [comment]
It looks quite good to me!

**Comment by DamirPorobic** [comment]
I'm done, give the CI some time to build, the feature should be available in latest continuous build.

**Comment by vaunchag** [comment]
Thank you very much!

I've tested it and looks good (although it took me a time to find where the new functionnality was). A couple of things:
- I think that modify Canvas should be considered like a "change", so that the file can be saved after it. Currently when changing the canvas the file cannot be saved with "only that change". One needs to do a further annotating action that triggers a change (an undo as a workaround if only the canvas change is desired).
- As per our recent comments on #91, I confirm that it looks strange not having the visual feedback of what is saved. It is odd the need to "have a look" each time to the "View > Modify Canvas" menu just to see/confirm what is the current canvas and what is going to be saved.

**Comment by DamirPorobic** [comment]
Good catch, forget about the first point. Will push a change in couple of minutes. 

Regarding second point, yes, I agree, we need to address that with #91. Let's see how that change improves the visibility and take it from there. I hope I can provide something to test by end of the week.

**Comment by DamirPorobic** [comment]
Changed pushed, give the CI some time to finish the build.

**Comment by vaunchag** [comment]
God, you're quick!
I'll test when available (your push is from 5h ago, but not yet available... not sure what is the normal time for the CI).
Regarding #91, no rush at all! I just gave the confirmation that, once tested as it is now, it looks unintuitive... but the functionality is already there and can at least be used. Great!

**Comment by DamirPorobic** [comment]
It was just a minor fix :)

The CI is slow lately, takes ~50min but you're right, can't see the build items. I have triggered it again. 

I'm working already on #91, I think it looks quite nice. Over that would come the canvas color, what ever you have picked there:

![image](https://user-images.githubusercontent.com/12796236/103412531-dd3dbe00-4b75-11eb-8fb0-5510153ffbf1.png)

**Comment by vaunchag** [comment]
Curiously, the new build isn't appearing... I'll wait for next year. :-)

Your proposal for #91 looks great! I assume that the checkboard represents transparency (as per usual convention).

**Comment by vaunchag** [comment]
Just confirming here that I also tested successfully this one. Many thanks again!

As a side note, for your information: I have recently recalled the editing interface of Shutter (the quite good Linux-only annotation tool now discontinued). The editor there constantly "offers" the canvas resize handlers permiting freely enlarging and reducing at any moment:

![image](https://user-images.githubusercontent.com/64960827/103695740-6fdac480-4f9d-11eb-9b0f-ebdaf6f74512.png)

I still find having this option quite convenient and "safe": the handlers are ofered and visible, but one does not have to use them if not interested in resizing the canvas.

Limitations there of things that may be often needed:

- Does not allow resizing the up or left sides.
- Does not permit setting the filling color (always transparent, in this case).

Both limitations are now covered by ksnip :-) ... although in a indirect modal window.
